api/base/views.py

```diff
@@ -437,7 +453,93 @@ def root(request, format=None, **kwargs):
     if utils.has_admin_scope(request):
         return_val['meta']['admin'] = True
 
-    return Response(return_val)
+    resp = Response(return_val)
+
+    # This is used current for our partnership with Sloan and can be deleted after their study is complete.
+    for key, value in cookies.items():
+        resp.set_cookie(key, value)
+
+    return resp
+
+
+def sloan_study_disambiguation(request):
+    """
+    This is a hack to set flags and cookies for out Sloan study, it can be deleted when the study is complete.
+    """
+    user = request.user
+    cookies = {}
+
+    flags = []
+    sloan_data = {}
+    for flag in Flag.objects.all():
+        active = flag.is_active(request._request)
+        if flag.name in SLOAN_FLAGS and flag.everyone is None:
+            sloan_data[flag.name] = active
+
+        # User tags should override any cookie info
+        if user and not user.is_anonymous:
+            if user.all_tags.filter(name=flag.name):
+                sloan_data[flag.name] = True
+                active = True
+            if user.all_tags.filter(name=f'no_{flag.name}'):
+                sloan_data[flag.name] = False
+                active = False
+
+        if active:
+            flags.append(flag.name)
+
+    referer_url = request.environ.get('HTTP_REFERER', '')
+    provider = get_provider_from_url(referer_url)
+
+    if provider and provider.in_sloan_study:
+        for key, value in sloan_data.items():
+            cookies.update(set_tags_and_cookies_for_sloan(user, key, value))
+
+    return flags, cookies
+
+
+def get_provider_from_url(referer_url: str) -> PreprintProvider:
+    """
+    Takes the many preprint refer urls and try to figure out the provider based on that.
+    This will be eliminated post-sloan.
+    :param referer_url:
+    :return: PreprintProvider
+    """
+
+    # matches custom domains:
+    provider_domains = list(PreprintProvider.objects.exclude(domain='').values_list('domain', flat=True))
+    provider_domains = [domains for domains in provider_domains if referer_url.startswith(domains)]
+    if provider_domains:
+        return PreprintProvider.objects.get(domain=provider_domains[0])
+
+    provider_ids_regex = '|'.join(list(PreprintProvider.objects.all().values_list('_id', flat=True)))
```

Comment:
use `re.escape` -- don't think we restrict what can be in a provider id, might be regex-meaningful
```suggestion
    provider_ids_regex = '|'.join([re.escape(id) for id in PreprintProvider.objects.all().values_list('_id', flat=True)])
```